met-api/src/met_api/models/membership.py

```diff
@@ -24,45 +24,80 @@ class Membership(BaseModel):
         """Get a survey."""
         memberships = db.session.query(Membership) \
-            .filter(Membership.engagement_id == engagement_id) \
+            .filter(and_(
+                Membership.engagement_id == engagement_id,
+                Membership.status == status,
+                bool(Membership.is_latest)
+            )) \
             .all()
         return memberships
 
     @classmethod
-    def find_by_user_id(cls, user_external_id) -> List[Membership]:
+    def find_by_user_id(cls, user_external_id, status=MembershipStatus.ACTIVE.value) -> List[Membership]:
         """Get memberships by user id."""
         memberships = db.session.query(Membership) \
             .join(StaffUser, StaffUser.id == Membership.user_id) \
             .filter(and_(StaffUser.external_id == user_external_id,
                          or_(Membership.type == MembershipType.TEAM_MEMBER,
-                             Membership.type == MembershipType.REVIEWER)
+                             Membership.type == MembershipType.REVIEWER),
+                         Membership.status == status,
+                         bool(Membership.is_latest)
                          )
                     ) \
             .all()
         return memberships
 
     @classmethod
     def find_by_engagement_and_user_id(cls, eng_id, userid, status=MembershipStatus.ACTIVE.value) \
-            -> List[Membership]:
+            -> Membership:
         """Get a survey."""
         memberships = db.session.query(Membership) \
             .join(StaffUser, StaffUser.id == Membership.user_id) \
             .filter(and_(Membership.engagement_id == eng_id,
                          Membership.user_id == userid,
-                         Membership.status == status
+                         Membership.status == status,
+                         bool(Membership.is_latest)
                          )
                     ) \
-            .all()
+            .first()
         return memberships
+
+    @classmethod
+    def create_new_version(cls, engagement_id, user_id, new_membership: dict) -> Membership:
+        """Create new version of membership."""
+        latest_membership = db.session.query(Membership) \
+            .filter(and_(Membership.engagement_id == engagement_id,
+                         Membership.user_id == user_id,
+                         bool(Membership.is_latest)
+                         )
+                    ) \
+            .first()
+        latest_membership.is_latest = False
+        latest_membership.save()
+
+        new_membership: Membership = Membership(
+            engagement_id=engagement_id,
+            user_id=user_id,
+            status=new_membership.get('status'),
+            type=new_membership.get('type'),
+            revoked_date=new_membership.get('revoked_date', None),
+            is_latest=True,
+            version=latest_membership.version + 1
+        )
+        new_membership.save()
+        return new_membership
```

Comment:
this is very clean..
the only improvement i can think is of using an update query directly and return a scalar.. that will avoid a full table fetch
https://stackoverflow.com/questions/45842715/how-to-use-returning-for-query-update-with-sqlalchemy

something like update(). returning(version)

But this code is clean, i almost dont want to change it at all.. , so i will the decision to you

Reply:
I can change it to that, it is not a very frequent call. I would value clarity and easiness to change for the maintainer here

Reply:
yea..i almost dont wanna touch this code,.